crates/rust-analyzer/src/main_loop.rs

```diff
@@ -428,18 +432,14 @@ fn loop_turn(
         && loop_state.in_flight_libraries == 0
     {
         loop_state.workspace_loaded = true;
-        let n_packages: usize = world_state.workspaces.iter().map(|it| it.n_packages()).sum();
-        if world_state.feature_flags.get("notifications.workspace-loaded") {
-            let msg = format!("workspace loaded, {} rust packages", n_packages);
-            show_message(req::MessageType::Info, msg, &connection.sender);
-        }
         world_state.check_watcher.update();
         pool.execute({
             let subs = loop_state.subscriptions.subscriptions();
             let snap = world_state.snapshot();
             move || snap.analysis().prime_caches(subs).unwrap_or_else(|_: Canceled| ())
         });
     }
+    send_startup_progress(&connection.sender, loop_state, world_state);
```

Comment:
Um.. so it will call in every `loop_turn` even `loop_state.workspace_loaded` is `true` and thus `WorkDoneProgressEnd` will send multiple times, right ?

Reply:
The `send_startup_progress` method only triggers when there's a change in `world_state.roots_to_scan`. So the `WorkDoneProgressEnd` will only send once.

Reply:
Oh I see ([this line](https://github.com/rust-analyzer/rust-analyzer/blob/5a1e531176bf6dab75e6c33cc30be6c5bb3ff345/crates/rust-analyzer/src/main_loop.rs#L716)), but how about `world_state.roots_to_scan` changed **after** `loop_state.workspace_loaded` is true ? Currently this invariant (`world_state.roots_to_scan` won't change after workspace loaded) is valid unintentionally, and we might support adding new vfs root on demand. (https://github.com/rust-analyzer/rust-analyzer/pull/3582#issuecomment-599243670). 

But anyway, I am happy if we just add a FIXME here to indicate this invariant has to be valid. :)

Reply:
Okay, I wasn't sure if the `roots_to_scan` was able to change after. I updated to make the code more clear and rely on the `workspace_loaded` check instead.